includes/EnvoimoinscherModel.php

```diff
@@ -1632,4 +1856,193 @@ public static function getNameCategory($id_eca)
+					break;
+			}
+		}
+		else
+		{
+			$error_message = $emc->l('Order not found');
+			$return = false;
+		}
+
+		// The push request has not been done correctly
+		if ($return === false)
+		{
+			$ip_address = $emc->l('Unknown address');
+			if (preg_match('/^([A-Za-z0-9.]+)$/', Tools::getRemoteAddr()))
+				$ip_address = Tools::getRemoteAddr();
+			$error_msg = sprintf(
+				$emc->l('%s. Target order : %s.Caller IP address : %s'),
+				$error_message,
+				$order,
+				$ip_address
+			);
+			Logger::addLog('[ENVOIMOINSCHER]['.time().'] '.$error_msg, 4, 1);
+		}
+		return $return;
+	}
+
+	public function updateTracking($order, $text, $localisation, $state, $date)
+	{
+		$emc = new Envoimoinscher();
+		$cookie = $emc->getContext()->cookie;
+
+		// Get module tracking configs
+		$confs = Configuration::getMultiple(array('EMC_ANN', 'EMC_ENVO', 'EMC_CMD', 'EMC_LIV'));
+
+		// Get the new state
+		$new_state = 0;
+		switch ($state)
+		{
+			case 'CMD':
+				$new_state = $confs['EMC_CMD'];
+			break;
+			case 'ENV':
+				$new_state = $confs['EMC_ENVO'];
+			break;
+			case 'ANN':
+				$message = new Message();
+				$texte = $emc->l('EnvoiMoinsCher : Dispatch cancelled');
+				$message->message = htmlentities($texte, ENT_COMPAT, 'UTF-8');
+				$message->id_order = $order;
+				$message->private = 1;
+				$message->add();
+				$new_state = $confs['EMC_ANN'];
+			break;
+			case 'LIV':
+				$new_state = $confs['EMC_LIV'];
+			break;
+			default:
+				return false;
+		}
+
+		// Get the last order state (prevent repeat of the same state)
+		$history_row = Db::getInstance()->ExecuteS('SELECT * FROM '._DB_PREFIX_.'order_history
+			 WHERE id_order = '.$order.' ORDER BY id_order_history DESC LIMIT 1');
+
+		// The order has already been delivered, no need to add more tracking info
+		if ($history_row[0]['id_order_state'] == $confs['EMC_LIV'])
+			return true;
+
+		// Update the order's history
+		if ((int)$new_state > 0 && $new_state != $history_row[0]['id_order_state'])
+		{
+			$history = new OrderHistory();
+			$history->id_order = $order;
+			$history->changeIdOrderState($new_state, $order);
+			$history->id_employee = (int)$cookie->id_employee;
+			$history->addWithemail();
+		}
+
+		// Generate a tracking message if no one was given
+		if ($text == '')
+		{
+			$cmd_row = Db::getInstance()->ExecuteS('SELECT * FROM '._DB_PREFIX_.'order_state_lang
+				 WHERE id_order_state = '.(int)$new_state.' AND id_lang = '.(int)$cookie->id_lang);
+			$text = $emc->l('Order\'s state : ').$cmd_row[0]['name'];
+		}
+
+		// Generate the default date if no one was given
+		if ($date == false)
+			$date = time();
+
+		// Add tracking info
+		Db::getInstance()->Execute('INSERT INTO '._DB_PREFIX_.'emc_tracking
+			 ('._DB_PREFIX_.'orders_id_order, state_et, date_et, text_et, localisation_et)
+			 VALUES ('.$order.', "'.$state.'", "'.date('Y-m-d H:i:s', $date).'", "'.pSQL($text).'", "'.pSQL($localisation).'")');
```

Comment:
It would be good to escape the data inserted into the database with pSQL or an int cast.